sklearn/utils/stats.py

```diff
@@ -1,11 +1,150 @@
+from collections.abc import Iterable
+
 import numpy as np
 
 from .extmath import stable_cumsum
 from .fixes import _take_along_axis
 
 
-def _weighted_percentile(array, sample_weight, percentile=50):
-    """Compute weighted percentile
+def _squeeze_arr(arr, n_dim):
+    return arr[0] if n_dim == 1 else arr
+
+
+def _nearest_rank(
+    array, sorted_idx, percentile, cum_weights, n_rows, n_cols, n_dim,
+):
+    """Compute weighted percentile using the nearest-rank approach.
+
+    Percentile can be computed using the nearest-rank:
+    https://en.wikipedia.org/wiki/Percentile
+
+    n = ceil(P / 100 * N)
+
+    where n is the index of the percentile value of the percentile P and N is
+    the number of samples.
+
+    This definition can be adapted with weights:
+
+    n_w = ceil(P / 100 * S_w) where S_w is the sum of the weights.
+    """
+    adjusted_percentile = percentile * cum_weights[-1]
+    percentile_idx = np.array([
+        np.searchsorted(cum_weights[:, i], adjusted_percentile[i])
+        for i in range(cum_weights.shape[1])
+    ])
+    percentile_idx = np.array(percentile_idx)
+    # In rare cases, percentile_idx equals to sorted_idx.shape[0]
+    max_idx = n_rows - 1
+    percentile_idx = np.apply_along_axis(
+        lambda x: np.clip(x, 0, max_idx), axis=0, arr=percentile_idx
+    )
+
+    col_index = np.arange(n_cols)
+    percentile_in_sorted = sorted_idx[percentile_idx, col_index]
+    percentile_value = array[percentile_in_sorted, col_index]
+    return _squeeze_arr(percentile_value, n_dim)
+
+
+def _interpolation_closest_ranks(
+    array, sorted_idx, sample_weight, sorted_weights, percentile, cum_weights,
+    interpolation, n_rows, n_cols, n_dim,
+):
+    """Compute the weighted percentile by interpolating between the adjacent
+    ranks.
+    Percentile can be computed with 3 different alternatives when using the
+    interpolation between adjacent ranks:
+    https://en.wikipedia.org/wiki/Percentile
+
+    These 3 alternatives depend of the value of a parameter C. NumPy uses
+    the variant where C=0 which allows to obtain a strictly monotonically
+    increasing function which is defined as:
```

Comment:
Actually I think numpy is the `C=1` variant. Wiki: https://en.wikipedia.org/wiki/Percentile#Second_variant,_%7F'%22%60UNIQ--postMath-00000047-QINU%60%22'%7F

Reply:
True

Reply:
With `C=0` I think you have the problem of 'ignoring' the first weight (instead of the last). So it's really tricky problem...